logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java

```diff
@@ -76,31 +76,53 @@ public WorkerLoop(
     public void run() {
         try {
             boolean isShutdown = false;
+            boolean isNackBatch = false;
             do {
                 isShutdown = isShutdown || shutdownRequested.get();
                 final QueueBatch batch = readClient.readBatch();
                 final boolean isFlush = flushRequested.compareAndSet(true, false);
                 if (batch.filteredSize() > 0 || isFlush) {
                     consumedCounter.add(batch.filteredSize());
-                    execution.compute(batch, isFlush, false);
-                    filteredCounter.add(batch.filteredSize());
-                    readClient.closeBatch(batch);
-
-                    if (isFlush) {
-                        flushing.set(false);
+                    isNackBatch = abortableCompute(batch, isFlush, false);
+                    if (!isNackBatch) {
+                        filteredCounter.add(batch.filteredSize());
+                        readClient.closeBatch(batch);
+
+                        if (isFlush) {
+                            flushing.set(false);
+                        }
                     }
                 }
-            } while (!isShutdown || isDraining());
-            //we are shutting down, queue is drained if it was required, now  perform a final flush.
-            //for this we need to create a new empty batch to contain the final flushed events
-            final QueueBatch batch = readClient.newBatch();
-            execution.compute(batch, true, true);
-            readClient.closeBatch(batch);
+            } while ((!isShutdown || isDraining()) && !isNackBatch);
+
```

Comment:
Is there ever going to be a situation where a nack happens _outside_ of worker shutdown, or is a nack expected to effectively shut down the worker?

Reply:
The nack is expected to interrupt the worker loop and do not close the batch. In such case any other batch shouldn't be processed, that's why of the following 
```java
if (!isNackBatch) {
  // consume last batch
}
```